pyro/distributions/hmm.py

```diff
@@ -0,0 +1,120 @@
+        conditional distributions between latent states. Should have rightmost
+        shape ``(num_steps, state_dim, state_dim)`` (time, old, new), and be
+        broadcastable to ``batch_shape + (num_steps, state_dim, state_dim)``.
+    :param torch.distriburtions.Distribution observation_dist: A conditional
+        distribution of observed data conditioned on latent state. The
+        ``.batch_shape`` should have rightmost size ``state_dim`` and be
+        broadcastable to ``batch_shape + (num_steps + state_dim)``. The
+        ``.event_shape`` may be arbitrary.
+    """
+    arg_constraints = {"initial_logits": constraints.real,
+                       "transition_logits": constraints.real}
+
+    def __init__(self, initial_logits, transition_logits, observation_dist, validate_args=None):
+        if initial_logits.dim() < 1:
+            raise ValueError("expected initial_logits to have at least one dim, "
+                             "actual shape = {}".format(initial_logits.shape))
+        if transition_logits.dim() < 2:
+            raise ValueError("expected transition_logits to have at least two dims, "
+                             "actual shape = {}".format(transition_logits.shape))
+        if len(observation_dist.batch_shape) < 1:
+            raise ValueError("expected observation_dist to have at least one batch dim, "
+                             "actual .batch_shape = {}".format(observation_dist.batch_shape))
+        time_shape = broadcast_shape(transition_logits.shape[-3:-2],
+                                     observation_dist.batch_shape[-2:-1])
+        if not time_shape:
+            raise ValueError("Expected either transition_logits or observation_dist to be time-dependent, "
+                             "actual shapes: {}, {}".format(transition_logits.shape,
+                                                            observation_dist.batch_shape))
+        event_shape = time_shape + observation_dist.event_shape
+        batch_shape = broadcast_shape(initial_logits.shape[:-1],
+                                      transition_logits.shape[:-3],
+                                      observation_dist.batch_shape[:-2])
+        self.initial_logits = initial_logits - initial_logits.logsumexp(-1, True)
+        self.transition_logits = transition_logits - transition_logits.logsumexp(-1, True)
+        self.observation_dist = observation_dist
+        super(DiscreteHMM, self).__init__(batch_shape, event_shape, validate_args=validate_args)
+
+    def expand(self, batch_shape, _instance=None):
+        new = self._get_checked_instance(DiscreteHMM, _instance)
+        batch_shape = torch.Size(broadcast_shape(self.batch_shape, batch_shape))
+        new.initial_logits = self.initial_logits.expand(batch_shape + (-1,))  # cheapest to expand
+        new.transition_logits = self.transition_logits
+        new.observation_dist = self.observation_dist
```

Comment:
why do `transition_logits` and `observation_dist` not get expanded?

Reply:
I've added the following comment in the code:

> We only need to expand one of the inputs, since batch_shape is determined
  by broadcasting all three. To save computation in _sequential_logmatmulexp(),
  we expand only initial_logits, which is applied only after the logmatmulexp.
  This is similar to the `._unbroadcasted_*` pattern used elsewhere in distributions.